libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -167,6 +169,228 @@ _rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux)
+
+void
+_rtld_sighandler(int sig, siginfo_t *info, void *_ucp)
+{
+	uintptr_t csp, rcsp;
+	ucontext_t *ucp = _ucp;
+
+	csp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("mrs	%0, rcsp_el0" : "=C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = rcsp;
+
+	thr_sighandler(sig, info, ucp);
+
+	rcsp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("msr	rcsp_el0, %0" :: "C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = csp;
+}
+
+void
+_rtld_sighandler_init(void *p)
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_sighandler);
+	thr_sighandler = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void *
+partition(struct tramp **inout, size_t len)
+{
+	void *hi = *inout;
+	void *lo = cheri_setbounds(hi, len);
+	ptraddr_t top, alignment;
+	size_t rem;
+
+	*inout = lo;
+	top = cheri_gettop(lo);
+	rem = cheri_gettop(hi) - top;
+	alignment = CHERI_REPRESENTABLE_ALIGNMENT(rem);
+	hi = cheri_setaddress(hi, __align_up(top, alignment));
+	return cheri_setbounds(hi, __align_down(rem, alignment));
+}
+
+static int
+tramp_pg_create(struct tramp_pg **out)
+{
+	struct tramp_pg *p;
+	p = mmap(NULL,
+		 getpagesize(),
+		 PROT_READ | PROT_WRITE | PROT_EXEC,
+		 MAP_ANON | MAP_PRIVATE,
+		 -1, 0);
+	if (p == MAP_FAILED)
+		rtld_die();
+	p->cursor = p->trampolines;
+	*out = p;
+	return 0;
+}
+
+void *
+tramp_pgs_append(const void *target, const Obj_Entry *dst)
+{
+	static struct tramp_pgs pgs = SLIST_HEAD_INITIALIZER(pgs);
+
+	extern const struct tramp __start_tramp_template_exe;
+	static const struct tramp *template_exe = &__start_tramp_template_exe;
+
+	extern const struct tramp __start_tramp_template_res;
+	static const struct tramp *templte_res = &__start_tramp_template_res;
+
+	const struct tramp *template;
+	struct tramp *t;
+	size_t len;
+
+	int n_retry = 0;
+	struct tramp_pg *pg = SLIST_FIRST(&pgs);
+	goto start;
+
+retry:
+	if (n_retry++)
+		rtld_die();
+	if (tramp_pg_create(&pg))
+		rtld_die();
+	SLIST_INSERT_HEAD(&pgs, pg, entries);
+
+start:
+	if (!pg)
+		goto retry;
+
+	if (cheri_getperm(target) & CHERI_PERM_EXECUTIVE)
+		template = template_exe;
+	else
+		template = templte_res;
+	len = cheri_getlen(template);
+
+	t = __align_up(pg->cursor, _Alignof(typeof(*t)));
+	pg->cursor = partition(&t, len);
+	if (!cheri_gettag(t))
+		goto retry;
+
+	memcpy(t, template, len);
+	// Arm64's i-cache and d-cache are not coherent. So we need to clean the d-cache
+	// and invalidate the i-cache. See https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/caches-and-self-modifying-code
+	// We derive the start and end addresses from pg so that they are aligned to the
+	// cache-line size, since pg is page aligned.
+	__clear_cache(cheri_copyaddress(pg, t->code),
+		      cheri_copyaddress(pg, (uintptr_t)t + len));
+
+	t->target = target;
+	if (dst)
+		t->compart_id = dst->compart_id;
+
+	t = cheri_clearperm(t, FUNC_PTR_REMOVE_PERMS);
+	return cheri_sealentry(t->code);
+}
+#endif
```

Comment:
At the start you have a blank line after the `#if` but this has no blank line before the `#endif`. They should be balanced.